benchmark/benchmark.py

```diff
@@ -0,0 +1,156 @@
+
+    print('{} loops, {} runs'.format(number, repeat))
+    res = timeit.repeat(func, number=number, repeat=repeat, globals=globals())
+    print("Timing per loop:", np.array(res) / number)
+
+
+def plot(title, kernels, labels, n_vals):
+    """Plot timings as a function of the parameter n.
+
+    Args:
+        title (str): plot title
+        kernels (Sequence[Callable[[Any], Any]]): parametrized benchmarks to time
+        labels (Sequence[str]): names of the kernels
+        n_vals (Sequence[Any]): values the benchmark parameter n takes
+    """
+    import perfplot
+
+    perfplot.show(
+        setup=lambda n: n,
+        kernels=kernels,
+        labels=labels,
+        n_range=n_vals,
+        xlabel="n",
+        title=title,
+    )
+
+
+def profile(func, identifier, *, min_time=5):
+    """Profile the given function.
+
+    Args:
+        func (Callable[[], Any]): function to profile
+        identifier (str): identifying part of the name of the file containing the results
+        min_time (float): func is called repeatedly until at least this many seconds have elapsed
+    """
+    import cProfile
+    import pstats
+
+    print("Minimum duration: {} seconds.".format(min_time))
+    pr = cProfile.Profile()
+    pr.enable()
+
+    t0 = time.process_time()
+    repeats = 0
+    while True:
+        func()
+        repeats += 1
+        elapsed = time.process_time() - t0
+        if elapsed > min_time:
+            break
+
+    pr.disable()
+    pr.dump_stats("pennylane_{}.pstats".format(identifier))
+    ps = pstats.Stats(pr).strip_dirs().sort_stats("tottime")  # "cumulative")
+    ps.print_stats()
+    print("{} repeats.".format(repeats))
+
+
+def cli():
+    """Parse the command line arguments, perform the requested action.
+    """
+    parser = argparse.ArgumentParser(description='PennyLane benchmarking tool')
+    parser.add_argument('--version', action='version', version=__version__)
+    parser.add_argument('-v', '--verbose', action='store_true', help='verbose mode')
+    parser.add_argument('-d', '--device', type=lambda x: x.split(','), default=['default.qubit'], help='comma-separated list of devices to run the benchmark on (default: %(default)s)')
+    parser.add_argument('-w', '--wires', type=int, default=3, help='number of wires to run the benchmark on (default: %(default)s)')
+    parser.add_argument('cmd', choices=['time', 'plot', 'profile'], help='function to perform')
+    parser.add_argument('benchmark', help='benchmark module')
+
+    args = parser.parse_args()
+
+    # look up information about the current HEAD Git commit
+    res = subprocess.run(
+        ["git", "log", "-1", "--pretty=%h %s"], stdout=subprocess.PIPE, encoding="utf-8", check=True
+    )
+    title = res.stdout
+    short_hash = title.split(" ", maxsplit=1)[0]
+
+    print("Benchmarking PennyLane", qml.version())
+    if args.verbose:
+        print("Verbose mode on, results may not be representative.")
+    print("Commit:", title)
```

Comment:
Should we also include the branch name here?

Reply:
I don't know. The branches could change but the commit hashes are forever.

Reply:
I imagine the one use case would be if someone ran the benchmark unknowingly on the wrong branch